File: Swift/Playgrounds/Advent of Code 2023.swiftpm/Solutions/Day21.swift
import Foundation

class Day21: AoCSolution {
    override init() {
        super.init()
        self.day = 21
        self.name = "Step Counter"
        self.emptyLinesIndicateMultipleInputs = true
    }
    
    override func solve(_ input: AoCInput) -> AoCResult {
        super.solve(input)
        
        let grid = AoCGrid2D(defaultValue: "#", rule: .rook)
        grid.load(data: input.textLines)
        let start = grid.getCoords(withValue: "S").first!
        grid.setValue(".", at: start)
        
        let part1 = solvePartOne(in: grid, at: start)
        let part2 = solvePartTwo(in: grid, at: start)
        
        return AoCResult(part1: "\(part1)", part2: "\(part2)")
    }
    
    func solvePartOne(in grid: AoCGrid2D, at start: AoCCoord2D) -> Int {
        let nSteps = grid.extent!.height < 15 ? 6 : 64 // Test (6) Challenge (64)
        
        var toVisit = [AoCCoord2D]()
        var nextSteps = Set<AoCCoord2D>()
        grid.neighbourCoords(at: start, withValue: ".")
            .forEach({nextSteps.insert($0)})
        
        for _ in 2...nSteps {
            toVisit = Array(nextSteps)
            //            var debugMarkers = Dictionary<AoCCoord2D,String>()
            //            toVisit.forEach({debugMarkers[$0] = "O"})
            //            grid.draw(markers: debugMarkers)
            nextSteps.removeAll()
            for plot in toVisit {
                grid.neighbourCoords(at: plot, withValue: ".")
                    .forEach {nextSteps.insert($0)}
            }
        }
        
        return nextSteps.count
    }
    
    func solvePartTwo(in grid: AoCGrid2D, at start: AoCCoord2D) -> Int {
        let ext = grid.extent!
        if ext.width < 131 { return -1 } // This code won't work on the test data
        
        // Do a BFS to create a grid of distances from start for all plots
        var toVisit = [AoCCoord2D]()
        var visited = Set<AoCCoord2D>([start])
        var nextSteps = Set<AoCCoord2D>()
        
        let distanceGrid = AoCGrid2D(defaultValue: " ", rule: .rook)
        distanceGrid.setValue(0, at: start)
        
        grid.neighbourCoords(at: start, withValue: ".")
            .filter({!visited.contains($0)})
            .forEach({nextSteps.insert($0)})
        
        var distance = 1
        while !nextSteps.isEmpty {
            toVisit = Array(nextSteps)
            nextSteps.removeAll()
            for plot in toVisit {
                distanceGrid.setValue(distance, at: plot)
                visited.insert(plot)
                grid.neighbourCoords(at: plot, withValue: ".")
                    .filter({!visited.contains($0)})
                    .forEach {nextSteps.insert($0)}
            }
            distance += 1
        }
        
        //https://github.com/villuna/aoc23/wiki/A-Geometric-solution-to-advent-of-code-2023,-day-21
        let even_corners = distanceGrid.values.map({$0 as! Int})
            .filter({$0 % 2 == 0 && $0 > 65}).count
        let odd_corners = distanceGrid.values.map({$0 as! Int})
            .filter({$0 % 2 == 1 && $0 > 65}).count
        
        let even_full = distanceGrid.values.map({$0 as! Int})
            .filter({$0 % 2 == 0}).count
        let odd_full = distanceGrid.values.map({$0 as! Int})
            .filter({$0 % 2 == 1}).count
        
        let n = (26501365 - ext.width / 2) / ext.width
        assert(n == 202300)
        
        let answer = ((n+1)*(n+1)) * odd_full + (n*n) * even_full
            - (n+1) * odd_corners + n * even_corners;
        
        return answer
    }
}

File: Perl/day21.pl
#!/usr/bin/env perl
use v5.42;

our $directory;
BEGIN { use Cwd; $directory = cwd; }
use lib $directory . '/lib';

use feature 'signatures';

use AOC::Util;
use AOC::Geometry;
use AOC::Grid;

my $DAY = sprintf("%02d", 21);
my $INPUT = $ARGV[0];
my $INPUT_FILE = 'day' . $DAY . '_' . $INPUT . '.txt';

my @input = read_input("../Input/$INPUT_FILE");

say "Advent of Code 2023, Day 21: Step Counter";

my $grid = g2_make('@');
$grid->load(@input);
my @s = $grid->coords("S");
$grid->set($s[0], '.');

solve_part_one($grid, $s[0]);
solve_part_two($grid, $s[0]);

exit( 0 );

sub solve_part_one($grid, $start) {
	my $iterations = ($grid->extent()->width < 20) ? 6 : 64;

	my %locations = ($start->to_str() => 1);
	for my $i (1..$iterations) {
		my %next_locations = ();

		for my $plot_str (keys %locations) {
			my $plot = c2_from_str($plot_str);
			my @n = grep {$grid->get_scalar($_) eq '.'} $grid->neighbors($plot);
			for my $n (@n) {
				$next_locations{$n->to_str()} = 'O';
			}
		}

		%locations = %next_locations;
	}

	#$grid->print(\%locations);

	say "Part One: the number of visited plots is " . scalar %locations;
}

sub solve_part_two($grid, $start) {
	my %visited = ();
	my %next_steps = ();
	my $ext = $grid->extent();

	my $distance_grid = g2_make();
	$distance_grid->set($start, 0);

	my @n = grep {$grid->get_scalar($_) eq '.'} $grid->neighbors($start);
	for my $n (@n) { my $s = $n->to_str(); if (!defined $visited{$s}) {$next_steps{$s} = 1 } }

	my $distance = 1;
	while (scalar %next_steps > 0) {
		my @to_visit = map {c2_from_str($_)} keys %next_steps;
		%next_steps = ();
		for my $plot (@to_visit) {
			$distance_grid->set($plot, $distance);
			$visited{$plot->to_str()} = 1;
			@n = grep {$ext->contains($_) && $grid->get_scalar($_) eq '.'} $grid->neighbors($plot);
			for my $n (@n) {
				my $s = $n->to_str();
				if (!defined $visited{$s}) {
					$next_steps{$s} = 1
				}
			}
		}
		$distance++;
	}

	# https://github.com/villuna/aoc23/wiki/A-Geometric-solution-to-advent-of-code-2023,-day-21
	my $hist_ref = $distance_grid->histogram();
	my %hist = %{$hist_ref};
	my $even_corners = 0;	my $odd_corners = 0;
	my $even_full = 0;		my $odd_full = 0;
	for my $value (keys %hist) {
		$even_corners += $hist{$value} if $value % 2 == 0 && $value > 65;
		$odd_corners += $hist{$value} if $value % 2 == 1 && $value > 65;

		$even_full += $hist{$value} if $value % 2 == 0;
		$odd_full += $hist{$value} if $value % 2 == 1;
	}

	my $w = $ext->width();
	my $n = sprintf("%.0f", ((26501365 - $w / 2) / $w));
	die "$n" unless ($n == 202300);

	my $answer = (($n+1)*($n+1)) * $odd_full + ($n*$n) * $even_full - ($n+1) * $odd_corners + $n * $even_corners;

	say "Part Two: the number of visited plots on an infinite grid is $answer.";
}
